### PR TITLE
Add Scalar Symbol

### PR DESCRIPTION
**Pull request by yajiedesign**
Add Scalar Symbol
https://github.com/dmlc/mxnet/issues/473

### CODE REVIEW

**Comment by antinucleon** [comment]
The backward dependency is incorrect. I will make more comment after I can
reach a computer

**Comment by tqchen** [comment]
Actually, what I mean by ScalarOp was ElementwiseBinaryScalarOp. Something like AddSymbolScalar, MinusSymbolScalar. So the operator can pass shape inference check, and we do not need to create a real NDArray for the scalars

**Comment by yajiedesign** [comment]
@tqchen I know

**Comment by yajiedesign** [comment]
@tqchen update

**Comment by yajiedesign** [comment]
I'm not sure Backward is right.please check.

**Comment by tqchen** [comment]
@yajiedesign can you add two testcases into https://github.com/dmlc/mxnet/blob/master/tests/python/unittest/test_operator.py Thanks!

File: src/operator/elementwise_binary_scalar_op-inl.h [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+}  // elembinary
+
+struct ScalarOpParam : public dmlc::Parameter<ScalarOpParam> {
+    // use int for enumeration
+    float value;
+    bool right;
+    DMLC_DECLARE_PARAMETER(ScalarOpParam) {
+        DMLC_DECLARE_FIELD(value)
+            .set_default(0)
+            .describe("Activation function to be applied.");
```
Comment (tqchen) [inline src/operator/elementwise_binary_scalar_op-inl.h]: change the decription to scalar value

File: src/operator/elementwise_binary_scalar_op-inl.h [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+enum ElementwiseBinaryScalarOpType {kPlus, kMinus, kMul, kDiv};
+}  // elembinary
+
+struct ScalarOpParam : public dmlc::Parameter<ScalarOpParam> {
+    // use int for enumeration
+    float value;
+    bool right;
+    DMLC_DECLARE_PARAMETER(ScalarOpParam) {
+        DMLC_DECLARE_FIELD(value)
+            .set_default(0)
```
Comment (tqchen) [inline src/operator/elementwise_binary_scalar_op-inl.h]: no default is needed

File: src/operator/elementwise_binary_scalar_op-inl.h [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+
+namespace elembinary {
+enum ElementwiseBinaryScalarOpInputs {kLhs};
+enum ElementwiseBinaryScalarOpOutputs {kOut};
+enum ElementwiseBinaryScalarOpType {kPlus, kMinus, kMul, kDiv};
+}  // elembinary
+
+struct ScalarOpParam : public dmlc::Parameter<ScalarOpParam> {
+    // use int for enumeration
+    float value;
```
Comment (tqchen) [inline src/operator/elementwise_binary_scalar_op-inl.h]: rename to scalar

File: src/operator/elementwise_binary_scalar_op-inl.h [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+    bool right;
+    DMLC_DECLARE_PARAMETER(ScalarOpParam) {
+        DMLC_DECLARE_FIELD(value)
+            .set_default(0)
+            .describe("Activation function to be applied.");
+        DMLC_DECLARE_FIELD(right)
```
Comment (tqchen) [inline src/operator/elementwise_binary_scalar_op-inl.h]: rename to scalar_on_right

File: src/operator/elementwise_binary_scalar_op-inl.h [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+            .set_default(false)
+            .describe("right");
```
Comment (tqchen) [inline src/operator/elementwise_binary_scalar_op-inl.h]: scalar operand is on the right

File: src/operator/elementwise_binary_scalar_op-inl.h [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+                       const std::vector<TBlob> &aux_args) {
+    using namespace mshadow;
+    using namespace mshadow::expr;
+    CHECK_EQ(in_data.size(), 1);
+    CHECK_EQ(out_data.size(), 1);
+    Stream<xpu> *s = ctx.get_stream<xpu>();
+    Tensor<xpu, 2> lhs = in_data[elembinary::kLhs].FlatTo2D<xpu, real_t>(s);
+    Tensor<xpu, 2> out = out_data[elembinary::kOut].FlatTo2D<xpu, real_t>(s);
+    if (!right_) {
+        Assign(out, req[elembinary::kOut], F<ForwardOp>(lhs, value_));
```
Comment (tqchen) [inline src/operator/elementwise_binary_scalar_op-inl.h]: two space

File: src/operator/elementwise_binary_scalar_op-inl.h [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+    Stream<xpu> *s = ctx.get_stream<xpu>();
+    Tensor<xpu, 2> m_out_grad = out_grad[elembinary::kOut].FlatTo2D<xpu, real_t>(s);
+    Tensor<xpu, 2> lhs_grad = in_grad[elembinary::kLhs].FlatTo2D<xpu, real_t>(s);
+    switch (GetOpType<ForwardOp>()) {
+      case elembinary::kPlus: {
+        Assign(lhs_grad, req[elembinary::kLhs], F<mshadow_op::identity>(m_out_grad));
+        break;
+      }
+      case elembinary::kMinus: {
+        Assign(lhs_grad, req[elembinary::kLhs], F<mshadow_op::identity>(m_out_grad));
```
Comment (tqchen) [inline src/operator/elementwise_binary_scalar_op-inl.h]: this depend on right, should be negation of gradient if it is right

File: src/operator/elementwise_binary_scalar_op-inl.h [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+    bool right_;
+};  // class ElementwiseBinaryScalarOp
+
+
+template<typename xpu>
+inline Operator* CreateElementwiseBinaryScalarOp_(elembinary::ElementwiseBinaryScalarOpType type,
+                                                  ScalarOpParam param) {
+  switch (type) {
+    case elembinary::kPlus:
+        return new ElementwiseBinaryScalarOp<xpu, mshadow::op::plus>(param);
```
Comment (tqchen) [inline src/operator/elementwise_binary_scalar_op-inl.h]: two soace

File: src/operator/elementwise_binary_scalar_op.cc [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+template<typename ForwardOp>
+Operator* ElementwiseBinaryScalarOpProp<ForwardOp>::CreateOperator(Context ctx) const {
+  DO_BIND_DISPATCH(CreateElementwiseBinaryScalarOp, GetOpType<ForwardOp>(), param_);
+}
+
+DMLC_REGISTER_PARAMETER(ScalarOpParam);
+
+MXNET_REGISTER_OP_PROPERTY(_PlusScalar, ElementwiseBinaryScalarOpProp<mshadow::op::plus>)
+.describe("Perform an elementwise plus.")
+.add_argument("lhs", "Symbol", "Input data to activation function.")
```
Comment (tqchen) [inline src/operator/elementwise_binary_scalar_op.cc]: consider change lhs to array, as it is the array input.

Change description to input array operand to the operation

File: tests/python/unittest/test_operator.py [hunk 10]
```diff
@@ -223,6 +223,21 @@ def test_swapaxes():
...
+    data_tmp = np.ones(shape)
+    data_tmp[:]=5
+    arr_data = mx.nd.array(data_tmp)
+    
+    test = (1+data+1)*2/5-0.2+3*data+5/data-data 
+    exe_test = test.bind(mx.cpu(), args=[arr_data])
+    exe_test.forward()
+    out = exe_test.outputs[0].asnumpy()
+    
+    npout = (1+data_tmp +1)*2/5-0.2+3*data_tmp+5/data_tmp-data_tmp
```
Comment (tqchen) [inline tests/python/unittest/test_operator.py]: can we add simple testcases for backward as well?

**Comment by tqchen** [comment]
Sorry for the delay. I made a pass on the code, can you update it accordingly with the comment? Thanks for the contribution!

**Comment by antinucleon** [comment]
LGTM

**Comment by pluskid** [comment]
Is it possible to add a `Constant` symbol that takes a constant scalar or ndarray and interact nicely with the symbolic API?

**Comment by rinuboney** [comment]
@pluskid that would be nice. I am trying to replicate the basic scalar multiplication tutorial from Theano ([https://github.com/Newmu/Theano-Tutorials/blob/master/0_multiply.py](https://github.com/Newmu/Theano-Tutorials/blob/master/0_multiply.py)) and TensorFlow ([https://github.com/nlintz/TensorFlow-Tutorials/blob/master/0_multiply.py](https://github.com/nlintz/TensorFlow-Tutorials/blob/master/0_multiply.py)) in MXNet. This works:

```python
import mxnet as mx

A = mx.sym.Variable('A')
B = mx.sym.Variable('B')

Y = A * B

multiply = Y.bind(ctx=mx.cpu(), args={'A': mx.nd.ones(1) * 1, 'B': mx.nd.ones(1) * 2})

multiply.forward()

print multiply.outputs[0].asscalar() #2

multiply = Y.bind(ctx=mx.cpu(), args={'A': mx.nd.ones(1) * 3, 'B': mx.nd.ones(1) * 3})

multiply.forward()

print multiply.outputs[0].asscalar() #9
```

I was expecting a more elegant solution (did I miss something?). I think a ```Constant``` symbol would allow that but I am not sure if it is actually useful since the ScalarOp is already included?

**Comment by pluskid** [comment]
Oh I was thinking about constant operator that could actually hold an ndarray and have a shape (other than just scalar).

**Comment by rinuboney** [comment]
Can you kindly explain how that is useful in comparison with ```Variable``` symbol?

**Comment by facundoq** [comment]
Sorry to necro this, but is this op documented? @pluskid the op you mention makes a lot of sense, is that implemented? up for a pr? i'm trying to debug some stuff in MXNet.jl and really need this.

**Comment by yajiedesign** [comment]
@facundoq This part has changed a lot.

**Comment by facundoq** [comment]
@yajiedesign I know I can declare a Variable and use mx.bind to set it's value, but that precludes using, for example, `fit`, because it doesn't accept an executor. Is there no way to easily do what @pluskid mentions?

**Comment by yajiedesign** [comment]
@facundoq  if you want calculation with scalar.you can direct use number. etc `otherSymbol * 5`.

**Comment by facundoq** [comment]
@yajiedesign I know, but sometimes you need an arbitrary matrix. 

For example, when using Spatial Transformer Networks, there's a subnet that computes an affine transform matrix. If I want to, say, force this matrix to be diagonal without touching the STN code, I could just do a broadcast_mul between the output of the subnet and a mask, [1 0 0 1 0 0]. 

With the symbolic API as it is I see no way to do this except using a Variable and bind.

**Pull request merged**